model/pdata/logs.go

```diff
@@ -34,6 +34,12 @@ type LogsUnmarshaler interface {
 	UnmarshalLogs(buf []byte) (Logs, error)
 }
 
+// LogsSizer returns the size of a Logs.
+type LogsSizer interface {
```

Comment:
@codeboten @kirbyquerby @tigrannajaryan

Should this be an "optional" interface that the Marshaler can implement or a separate interface?

Reply:
Here is just the comment, but look how it is used https://github.com/open-telemetry/opentelemetry-collector/pull/3818/files#diff-f8d7e57fb88bb32010db5d21c4aa5ad1feffe74790db7fd7c5c11c5ceec5c0c3R43

I was thinking that the `NewProtobufLogsMarshaler` will return a LogsMarshaler and users will try to cast to this "sizer" to see if that is supported instead of having separate "NewProtobufLogsSizer".

Reply:
Not sure if that is better, the only 3 new public funcs are the NewSizer methods so it may be good enough. Need some more thoughts here.

Reply:
One thing we could do is have the `*Sizer` interfaces extend the `*Marshaler` interfaces, which would imply that every `Sizer` is a `Marshaler`. Then we could alter the return signatures of `NewProtobuf*Marshaler` (and maybe the method names) to return `*Sizer`s instead. Other `Marshaler` implementations could then choose to either only implement`Marshaler` or also implement `Sizer`.

I'm personally not a huge fan of having users make runtime casts and would prefer that the underlying type be known at compile time. Another thing we could do is add `*Size()` to the `*Marshaler` interfaces + add an error to its return signature, and let implementations return an error if it's unsupported, which could be preferable to type assertions (but still not great imo).

Throwing out one more idea, we could have each marshaler implementation return its own interface which extends (composes? terminology is hard 🙃) all the interfaces it implements, eg:
```go
type ProtobufMarshalerSizer {
  pdata.TracesMarshaler
  pdata.MetricsMarshaler
  pdata.LogsMarshaler
  pdata.TracesSizer
  pdata.MetricsSizer
  pdata.LogsSizer
}

func NewProtobufMarshalerSizer() ProtobufMarshalerSizer {
  return newPbMarshaler()
}
```

This would let the supported features be clear at compile time while still allowing for flexibility in choosing what to implement.

Reply:
That will be very confusing from the "structure" perspective.

Reply:
To resolve this (postpone the decision), let's remove the New..Sizer funcs and in the batch processor do a New..Marshaler().(pdata.Sizer) conversion for the moment

Reply:
Added the casts for now + a TODO and created #3842 for the discussion on how we want to handle this.